src/torrent.cpp

```diff
@@ -1138,6 +1138,50 @@ bool is_downloading_state(int const st)
 		if (m_peer_list) m_peer_list->clear();
 	}
 
+	void torrent::set_sequential_range(piece_index_t first_piece, piece_index_t last_piece)
+	{
+		if (!has_picker()) {
+			if (!valid_metadata() || !m_connections_initialized) return;
+			else if (!m_have_all
+				|| settings().get_int(settings_pack::suggest_mode)
+				== settings_pack::suggest_read_cache)
+				need_picker();
+			else return;
+		}
+		m_sequential_download = true;
+		m_picker->set_sequential_range(first_piece, last_piece);
+	}
+
+	void torrent::set_sequential_start(piece_index_t first_piece)
+	{
+		if (!has_picker()) {
+			if (!valid_metadata() || !m_connections_initialized) return;
+			else if (!m_have_all
+				|| settings().get_int(settings_pack::suggest_mode)
+				== settings_pack::suggest_read_cache)
+				need_picker();
+			else return;
+		}
+		// m_torrent_file->files() is file_storage
+		// find the last file in piece `first_piece`
+		auto const f = m_torrent_file->files().map_block(first_piece, m_torrent_file->files().piece_size(first_piece) - 1, 1).back().file_index;
+
+		// find the last piece in file `f`
+		std::int64_t piece_size = m_torrent_file->files().piece_length();
+		auto const last_piece = piece_index_t(
+			std::lldiv(
+			(m_torrent_file->files().file_offset(f)
+			+
+			m_torrent_file->files().file_size(f)
+			-
+			1)
+			,///
+			piece_size
+			).quot
```

Comment:
Arvid either when using plain `/` or this C++ lldiv fn, I get warnings-as-errors on some CI build compilers, eg.:
```
error: implicit conversion loses integer precision: 'long long' to 'int' [-Werror,-Wshorten-64-to-32]
                        ).quot
```
I made sure both operands are 64 bits, so how come the `/` operator is implicitly converting them to 32 bits? Or why do I get this error?

Reply:
the underlying type of `piece_index_t` is `int`. It looks like you're doing math in `int64` and then assigning it into a `piece_index_t` (i.e. `int`). you need to explicitly cast it to `int` to avoid the warning (and be confident the result is never > `INT_MAX`)

Reply:
there are limits when loading torrents that ensures there cannot be more than `INT_MAX` pieces btw, so you're not on the hook for that last part.

Reply:
but, I think this function belongs as a member on `file_storage`. Partly to make it easier for clients to access in general, but also to allow it to be unit tested. maybe something like `file_storage::file_first_piece()` and `file_storage::file_last_piece()`

Reply:
> the underlying type of `piece_index_t` is `int`. It looks like you're doing math in `int64` and then assigning it into a `piece_index_t` (i.e. `int`). you need to explicitly cast it to `int` to avoid the warning

But isn't the result *already* being cast to a piece_index_t just by being within the `piece_index_t( )` ?

And also, from the compiler output, it specifically points to the `/` operator or `.quot` member as the culprit, as though - even with both operands being 64 bits - the result somehow being 32 bits is what it is complaining about as losing precision.